src/entity/object/AreaEffectCloud.php

```diff
@@ -0,0 +1,462 @@
+use pocketmine\network\mcpe\protocol\types\entity\EntityMetadataProperties;
+use pocketmine\utils\Binary;
+use pocketmine\world\particle\PotionSplashParticle;
+use function array_map;
+use function count;
+use function max;
+use function round;
+
+class AreaEffectCloud extends Entity{
+
+	public const DURATION = 600; // in ticks
+	public const DURATION_ON_USE = 0; // in ticks
+
+	public const WAIT_TIME = 10; // in ticks
+	public const REAPPLICATION_DELAY = 40; // in ticks
+
+	public const DEFAULT_RADIUS = 3.0; // in blocks
+	public const RADIUS_CHANGE_ON_PICKUP = -0.5; // in blocks
+	public const RADIUS_ON_USE = -0.5; // in blocks
+	public const RADIUS_PER_TICK = -0.005; // in blocks
+
+	public const TAG_POTION_ID = "PotionId"; //TAG_Short
+	protected const TAG_SPAWN_TICK = "SpawnTick"; //TAG_Long
+	protected const TAG_DURATION = "Duration"; //TAG_Int
+	protected const TAG_PICKUP_COUNT = "PickupCount"; //TAG_Int
+	protected const TAG_DURATION_ON_USE = "DurationOnUse"; //TAG_Int
+	protected const TAG_REAPPLICATION_DELAY = "ReapplicationDelay"; //TAG_Int
+	public const TAG_INITIAL_RADIUS = "InitialRadius"; //TAG_Float
+	protected const TAG_RADIUS = "Radius"; //TAG_Float
+	protected const TAG_RADIUS_CHANGE_ON_PICKUP = "RadiusChangeOnPickup"; //TAG_Float
+	protected const TAG_RADIUS_ON_USE = "RadiusOnUse"; //TAG_Float
+	protected const TAG_RADIUS_PER_TICK = "RadiusPerTick"; //TAG_Float
+	protected const TAG_EFFECTS = "mobEffects"; //TAG_List
+
+	public static function getNetworkTypeId() : string{ return EntityIds::AREA_EFFECT_CLOUD; }
+
+	protected int $age = 0;
+
+	protected EffectContainer $effectContainer;
+
+	/** @var array<int, int> entity ID => expiration */
+	protected array $victims = [];
+
+	protected int $waiting = self::WAIT_TIME;
+	protected int $duration = self::DURATION;
+	protected int $durationChangeOnUse = self::DURATION_ON_USE;
+
+	protected int $reapplicationDelay = self::REAPPLICATION_DELAY;
+
+	protected int $pickupCount = 0;
+	protected float $radiusChangeOnPickup = self::RADIUS_CHANGE_ON_PICKUP;
+
+	protected float $radius = self::DEFAULT_RADIUS;
+	protected float $radiusChangeOnUse = self::RADIUS_ON_USE;
+	protected float $radiusChangePerTick = self::RADIUS_PER_TICK;
+
+	//TODO: Custom particle.
+
+	public function __construct(
+		Location $location,
+		protected PotionType $potionType,
+		protected float $initialRadius = self::DEFAULT_RADIUS,
+		?CompoundTag $nbt = null
+	){
+		parent::__construct($location, $nbt);
+	}
+
+	protected function getInitialSizeInfo() : EntitySizeInfo{ return new EntitySizeInfo(0.5, $this->radius * 2); }
+
+	protected function getInitialDragMultiplier() : float{ return 0.0; }
+
+	protected function getInitialGravity() : float{ return 0.0; }
+
+	protected function initEntity(CompoundTag $nbt) : void{
+		parent::initEntity($nbt);
+
+		$this->effectContainer = new EffectContainer();
+		$this->effectContainer->setDefaultBubbleColor(PotionSplashParticle::DEFAULT_COLOR());
```

Comment:
As above, don't like this. Default colour of bubbles shouldn't be the EffectContainer's responsibility, since it's only going to be used by AreaEffectCloud.